File: dotnet/src/Envdoctor/Parsers/K8sParser.cs
using Envdoctor.Models;

namespace Envdoctor.Parsers;

/// Parser for Kubernetes manifests.
///
/// Matches YAML files that look like Kubernetes resources (have apiVersion and
/// kind). Extracts container environment definitions and `${VAR}` / `$VAR`
/// interpolations from command/args/env values.
public sealed class K8sParser : IParser
{
    public string Id => "kubernetes";

    public bool MatchPath(string filePath)
    {
        var ext = Path.GetExtension(filePath).TrimStart('.').ToLowerInvariant();
        return ext is "yaml" or "yml";
    }

    public EnvironmentFile Parse(string content, string filePath)
    {
        var docArray = YamlFacade.LoadAll(content);

        var variables = new List<EnvironmentVariable>();
        var usages = new List<EnvironmentVariable>();

        foreach (var doc in docArray)
        {
            if (!LooksLikeK8s(doc))
                continue;
            WalkResource((Dictionary<string, object?>)doc!, filePath, variables, usages);
        }

        return new EnvironmentFile
        {
            FilePath = filePath,
            Format = FileFormat.Kubernetes,
            Variables = EnvironmentVariable.Merge(variables),
            Usages = EnvironmentVariable.Merge(usages),
        };
    }

    private static bool LooksLikeK8s(object? doc) =>
        doc is Dictionary<string, object?> obj &&
        obj.TryGetValue("apiVersion", out var apiVersion) && apiVersion is string &&
        obj.TryGetValue("kind", out var kind) && kind is string;

    private static Origin OriginAt(string filePath, int? line, OriginKind kind = OriginKind.Definition) =>
        new()
        {
            FilePath = filePath,
            Line = line,
            Kind = kind,
            Format = OriginFormat.Kubernetes,
        };

    private static void WalkResource(
        Dictionary<string, object?> doc,
        string filePath,
        List<EnvironmentVariable> variables,
        List<EnvironmentVariable> usages)
    {
        var kind = doc.TryGetValue("kind", out var k) ? k as string : null;

        // ConfigMap data keys become definitions.
        if (kind == "ConfigMap")
        {
            if (GetObject(doc, "data") is { } data)
            {
                foreach (var (key, value) in data)
                {
                    if (value is not string valueStr || key.Length == 0)
                        continue;
                    variables.Add(EnvironmentVariable.Create(
                        key, valueStr, new List<Origin> { OriginAt(filePath, null) }));
                }
            }
            return;
        }

        var spec = GetObject(doc, "spec");
        if (spec is null)
            return;

        var template = GetObject(spec, "template");
        var podSpec = template is not null ? GetObject(template, "spec") : spec;
        if (podSpec is null)
            return;

        var containers = (GetArray(podSpec, "containers") ?? new List<object?>())
            .Concat(GetArray(podSpec, "initContainers") ?? new List<object?>());

        foreach (var container in containers)
        {
            if (container is not Dictionary<string, object?> containerObj)
                continue;

            foreach (var raw in GetArray(containerObj, "env") ?? new List<object?>())
            {
                if (raw is not Dictionary<string, object?> entry)
                    continue;
                if (entry.TryGetValue("name", out var nameObj) && nameObj is string name && name.Length > 0)
                {
                    if (entry.TryGetValue("value", out var value) && value is string valueStr)
                    {
                        variables.Add(EnvironmentVariable.Create(
                            name, valueStr, new List<Origin> { OriginAt(filePath, null) }));
                    }
                    else if (entry.ContainsKey("valueFrom"))
                    {
                        // Referenced but value provided elsewhere (ConfigMap/Secret).
                        usages.Add(EnvironmentVariable.Create(
                            name, null, new List<Origin> { OriginAt(filePath, null, OriginKind.Usage) }));
                    }
                }
            }

            foreach (var raw in GetArray(containerObj, "envFrom") ?? new List<object?>())
            {
                if (raw is not Dictionary<string, object?> entry)
                    continue;
                var prefix = entry.TryGetValue("prefix", out var p) && p is string ps ? ps : "";
                if (GetObject(entry, "configMapRef") is { } configMapRef &&
                    configMapRef.TryGetValue("name", out var cmName) && cmName is string)
                {
                    usages.Add(EnvironmentVariable.Create(
                        $"{prefix}*", null, new List<Origin> { OriginAt(filePath, null, OriginKind.Usage) }));
                }
                if (GetObject(entry, "secretRef") is { } secretRef &&
                    secretRef.TryGetValue("name", out var secName) && secName is string)
                {
                    usages.Add(EnvironmentVariable.Create(
                        $"{prefix}*", null, new List<Origin> { OriginAt(filePath, null, OriginKind.Usage) }));
                }
            }

            // Interpolations in command/args.
            foreach (var key in new[] { "command", "args" })
            {
                var list = GetArray(containerObj, key);
                if (list is null)
                    continue;
                foreach (var item in list)
                {
                    if (item is not string itemStr)
                        continue;
                    foreach (var interp in YamlInterp.ScanInterpolations(itemStr))
                    {
                        usages.Add(EnvironmentVariable.Create(
                            interp.Name, null,
                            new List<Origin> { OriginAt(filePath, interp.Line, OriginKind.Usage) }));
                    }
                }
            }
        }
    }

    private static Dictionary<string, object?>? GetObject(Dictionary<string, object?> obj, string key) =>
        obj.TryGetValue(key, out var value) && value is Dictionary<string, object?> map ? map : null;

    private static List<object?>? GetArray(Dictionary<string, object?> obj, string key) =>
        obj.TryGetValue(key, out var value) && value is List<object?> list ? list : null;
}

File: dart/lib/src/parsers/k8s_parser.dart
/// Parser for Kubernetes manifests.
///
/// Matches YAML files that look like Kubernetes resources (have apiVersion and
/// kind). Extracts container environment definitions and `${VAR}` / `$VAR`
/// interpolations from command/args/env values.
library;

import '../models/environment_file.dart';
import '../models/environment_variable.dart';
import '../models/origin.dart';
import '../utils/paths.dart';
import 'parser.dart';
import 'yaml_facade.dart';
import 'yaml_interp.dart';

class K8sParser implements Parser {
  @override
  String get id => 'kubernetes';

  @override
  bool matchPath(String filePath) {
    final ext = extensionOf(filePath).toLowerCase();
    return ext == 'yaml' || ext == 'yml';
  }

  @override
  EnvironmentFile parse(String content, String filePath) {
    final docArray = yamlLoadAll(content);

    final variables = <EnvironmentVariable>[];
    final usages = <EnvironmentVariable>[];

    for (final doc in docArray) {
      if (!_looksLikeK8s(doc)) continue;
      _walkResource(doc as Map<String, Object?>, filePath, variables, usages);
    }

    return EnvironmentFile(
      filePath: filePath,
      format: FileFormat.kubernetes,
      variables: EnvironmentVariable.merge(variables),
      usages: EnvironmentVariable.merge(usages),
    );
  }

  static bool _looksLikeK8s(Object? doc) =>
      doc is Map<String, Object?> &&
      doc['apiVersion'] is String &&
      doc['kind'] is String;

  static Origin _originAt(String filePath, int? line,
          {OriginKind kind = OriginKind.definition}) =>
      Origin(filePath: filePath, line: line, kind: kind, format: OriginFormat.kubernetes);

  static Map<String, Object?>? _getObject(Map<String, Object?> obj, String key) {
    final value = obj[key];
    return value is Map<String, Object?> ? value : null;
  }

  static List<Object?>? _getArray(Map<String, Object?> obj, String key) {
    final value = obj[key];
    return value is List<Object?> ? value : null;
  }

  static void _walkResource(Map<String, Object?> doc, String filePath,
      List<EnvironmentVariable> variables, List<EnvironmentVariable> usages) {
    final kind = doc['kind'] as String?;

    // ConfigMap data keys become definitions.
    if (kind == 'ConfigMap') {
      final data = _getObject(doc, 'data');
      if (data != null) {
        for (final entry in data.entries) {
          if (entry.value is! String || entry.key.isEmpty) continue;
          variables.add(EnvironmentVariable.create(
              entry.key, entry.value as String, [_originAt(filePath, null)]));
        }
      }
      return;
    }

    final spec = _getObject(doc, 'spec');
    if (spec == null) return;

    final template = _getObject(spec, 'template');
    final podSpec = template != null ? _getObject(template, 'spec') : spec;
    if (podSpec == null) return;

    final containers = <Object?>[
      ...?_getArray(podSpec, 'containers'),
      ...?_getArray(podSpec, 'initContainers'),
    ];

    for (final container in containers) {
      if (container is! Map<String, Object?>) continue;

      for (final raw in _getArray(container, 'env') ?? <Object?>[]) {
        if (raw is! Map<String, Object?>) continue;
        final nameObj = raw['name'];
        if (nameObj is! String || nameObj.isEmpty) continue;
        final value = raw['value'];
        if (value is String) {
          variables.add(EnvironmentVariable.create(
              nameObj, value, [_originAt(filePath, null)]));
        } else if (raw.containsKey('valueFrom')) {
          // Referenced but value provided elsewhere (ConfigMap/Secret).
          usages.add(EnvironmentVariable.create(
              nameObj, null, [_originAt(filePath, null, kind: OriginKind.usage)]));
        }
      }

      for (final raw in _getArray(container, 'envFrom') ?? <Object?>[]) {
        if (raw is! Map<String, Object?>) continue;
        final prefixObj = raw['prefix'];
        final prefix = prefixObj is String ? prefixObj : '';
        final configMapRef = _getObject(raw, 'configMapRef');
        if (configMapRef != null && configMapRef['name'] is String) {
          usages.add(EnvironmentVariable.create(
              '$prefix*', null, [_originAt(filePath, null, kind: OriginKind.usage)]));
        }
        final secretRef = _getObject(raw, 'secretRef');
        if (secretRef != null && secretRef['name'] is String) {
          usages.add(EnvironmentVariable.create(
              '$prefix*', null, [_originAt(filePath, null, kind: OriginKind.usage)]));
        }
      }

      // Interpolations in command/args.
      for (final key in ['command', 'args']) {
        final list = _getArray(container, key);
        if (list == null) continue;
        for (final item in list) {
          if (item is! String) continue;
          for (final interp in scanInterpolations(item)) {
            usages.add(EnvironmentVariable.create(
                interp.name, null, [_originAt(filePath, interp.line, kind: OriginKind.usage)]));
          }
        }
      }
    }
  }
}
